### PR TITLE
Addressing docker comments

### PR DESCRIPTION
**Pull request by bam241**
This address comments from #660 about Docker.

I added something from me as well, see if I can separate the housekeeping jobs from the others... and that way simply the travis CI file

### CODE REVIEW

**Comment by bam241** [comment]
@kkiesling @pshriwise @gonuke this is ready for review.

File: docker/build_geant4.sh
Comment (pshriwise) [inline docker/build_geant4.sh:10]: In #660 they mention verifying our file against a checksum. Where do we stand on adding that?

**Review by pshriwise — COMMENTED** [review]
Just one small comment from me. Nice cleanup of the housekeeping check!

File: docker/build_geant4.sh
Comment (bam241) [reply docker/build_geant4.sh:10]: did not found any checksum for geant4 tarball :(

File: docker/Dockerfile
Comment (ljacobson64) [inline docker/Dockerfile]: Why was sphinx moved out of this file?

File: docker/Dockerfile
Comment (ljacobson64) [inline docker/Dockerfile:17]: Why was astyle moved out of this file? Everything that could possibly mess with system libraries (i.e. apt or pip) was put in this file.

File: docker/build_dagmc.sh
Comment (ljacobson64) [inline docker/build_dagmc.sh:16]: The reason these lines were commented is that they would normally be part of a DAGMC build script, but in this specific case, we get the DAGMC source code by copying it into the docker image, not cloning it from github.

File: docker/build_dagmc.sh
Comment (ljacobson64) [inline docker/build_dagmc.sh:32]: The dagmc_build_dir would normally be deleted after building DAGMC, but we need to keep it in this specific case in order to run the tests.

File: .travis.yml
Comment (ljacobson64) [inline .travis.yml]: The housekeeping run should use 18.04

File: docker/Dockerfile
Comment (ljacobson64) [inline docker/Dockerfile:18]: Removing these lines will disallow the ability to do the housekeeping run on Ubuntu 16.04. This is fine, as long as the housekeeping run is changed back to being on 18.04.

**Review by ljacobson64 — CHANGES_REQUESTED** [review]

File: .travis.yml
Comment (bam241) [reply .travis.yml]: oups this was not intended

File: docker/build_dagmc.sh
Comment (kkiesling) [reply docker/build_dagmc.sh:16]: @ljacobson64 is there any reason to leave this line in as a comment w/ justification or can we just remove it all together? (see [here](https://github.com/svalinn/DAGMC/pull/660#discussion_r368664684))

File: docker/build_dagmc.sh
Comment (kkiesling) [reply docker/build_dagmc.sh:32]: My same question above applies here too- do you think it should be left in as a comment w/ an explanation or removed?

File: docker/Dockerfile
Comment (kkiesling) [reply docker/Dockerfile]: It looks like anything having to do w/ housekeeping was moved to the housekeeping.sh file.

**Review by kkiesling — APPROVED** [review]
I have no major comments. Thanks @bam241!

**Comment by bam241** [comment]
@ljacobson64 to avoid having the housekeeping stuff everywhere and only used for 1 job in ubuntu 18.04. I moved `astyle` and `sphinx` install in the housekeeping scripts.

File: docker/build_dagmc.sh
Comment (bam241) [reply docker/build_dagmc.sh:32]: @ljacobson64 

agreed with @kkiesling if this is commented without explanation it should not be part of the build.
I'd be happy to add them back with a line of explanation. 

when you mean run the tests, do you mean run the test locally ? 
Because we run the test in CI without those...

File: docker/build_dagmc.sh
Comment (bam241) [reply docker/build_dagmc.sh:16]: If I understand correctly, there are there just because the normal install `DAGMC` procedure should include them. But in this case as we copy the `DAGMC` from the PR (using `Travis`) we don't need them.

If this is correct, I don't think they should be there:
Those scripts are dedicated for CI not to show people how to install `DAGMC`.
It is probably a good idea to have a version of them somewhere to facilitate `DAGMC` install but for me this is definitely not the place.

Comment (kkiesling) [reply docker/build_dagmc.sh:16]: I agree w/ this assessment @bam241. I think the lines here should be removed and if we want a docker to help users, we need to host that elsewhere.

File: docker/build_geant4.sh
Comment (gonuke) [reply docker/build_geant4.sh:10]: We could generate a checksum against the version we are pulling

File: docker/travis.housekeeping.sh
Comment (gonuke) [inline docker/travis.housekeeping.sh]: Why are we moving this to here?

File: docker/travis.tests.sh
Comment (gonuke) [inline docker/travis.tests.sh:22]: The comment suggested streaming these to `stderr` and `stdout` - do we do this now?  Or is that default behavior?

**Comment by gonuke** [comment]
I'm ambivalent on the housekeeping changes.  They may happen fairly often, which means a lot of `apt install` and `pip installing` for stuff we could do once.  Should we have different docker images? Would that help?

File: docker/build_geant4.sh
Comment (bam241) [reply docker/build_geant4.sh:10]: make sense !

File: docker/travis.tests.sh
Comment (bam241) [reply docker/travis.tests.sh:22]: my understanding of the comment was the default behavior streams them to `stderr`

File: docker/build_dagmc.sh
Comment (ljacobson64) [reply docker/build_dagmc.sh:16]: For maintainability purposes, I think it would be good to have some indication of why the source code already exists in this directory. (The current situation with just a commented `git clone` line is also not very helpful.) Perhaps a comment line that says

```
# DAGMC source code already exists in this directory because it was copied into
# the docker image by travis.yml
```

File: docker/build_dagmc.sh
Comment (ljacobson64) [reply docker/build_dagmc.sh:32]: I think this line can be deleted.

To answer @bam241's question: I do mean for running the tests in CI. The CI runs `make test` in the DAGMC build directory, so the DAGMC build directory must exist in order to run the tests. See here: https://github.com/bam241/DAGMC/blob/fix_docker/docker/travis.tests.sh#L28

File: docker/build_dagmc.sh
Comment (bam241) [reply docker/build_dagmc.sh:16]: I think I understand,

the line is left there to highlight that we don't need to copy DAGMC source, while we normally would in a standard installation process.
So the line is basically there for educational purposes...
Is that correct @ljacobson64 ?

I would definitely prefer to not mix purposes in those scripts. So if there are used for CI install and testing, I would remove the git clone line and add the comment at the head of the file.

**Comment by bam241** [comment]
I was thinking @ljacobson64 @kkiesling @pshriwise @gonuke 
For separation of concern what do you think about:
- creating a `.travis_scripts` folder containing all the scripts required for CI (exept the `update_docker.sh` one)
- keep only the `Dockerfile` and the `update_docker.sh` macro in the `docker` folder
- create a `script` folder, with a version of the install scripts dedicated to installation for users ?

**Comment by ljacobson64** [comment]
That kind of sounds like overkill. I'd rather just delete those 2 lines we were discussing

**Comment by bam241** [comment]
> That kind of sounds like overkill. I'd rather just delete those 2 lines we were discussing

Yes but I feel like it will clarify the purpose of everything.
It is just that it does not seems obvious to me, that the scripts used to install and run the tests in CI would be located in the `docker` folder

**Comment by ljacobson64** [comment]
Would it make more sense if the `docker` folder were renamed `CI`?

**Comment by bam241** [comment]
@gonuke @kkiesling @ljacobson64 this should  be ready for an around of review

**Comment by bam241** [comment]
@ljacobson64 @kkiesling I think I am now satisfied with the way it looks. 

cc @pshriwise @gonuke

**Comment by gonuke** [comment]
Should we test some of this by adding an automatic build to dockerhub?

**Comment by ljacobson64** [comment]
How does that work?

**Comment by gonuke** [comment]
You point dockerhub at a Dockerfile in a github repo and everytime that repo changes, it rebuilds the image.  We have deployed it for 
* https://hub.docker.com/repository/docker/svalinn/pymoab-py2-18.04
* https://hub.docker.com/repository/docker/svalinn/pymoab-py2-16.04
* https://hub.docker.com/repository/docker/svalinn/pymoab-visit-py2-18.04
and it seems to work well.

So far they have all been for images whose configuration is 100% defined in the dockerfile, so I don't know how it will behave for this image that requires scripts...

**Comment by bam241** [comment]
@gonuke this seems like a good idea, will take a look at this

**Comment by gonuke** [comment]
I have something running here: https://hub.docker.com/repository/docker/svalinn/dagmc-ci-ubuntu-18.04/builds

**Comment by gonuke** [comment]
The dockerhub build is taking a long time....  probably not surprising

**Comment by gonuke** [comment]
Looks like it might have timed out...

**Comment by bam241** [comment]
@gonuek if we want to do the automated one, we will probably need 2 dockerfiles one for each ubuntu version

**Comment by ljacobson64** [comment]
We could stop supporting 16.04 since it's 4 years old now

**Comment by ljacobson64** [comment]
Although in that case we may as well start supporting 20.04...

**Comment by bam241** [comment]
@gonuke I think we will have problem....
timeout limits is set to 2h
https://success.docker.com/article/what-are-the-current-resource-limits-placed-on-automated-builds

**Comment by bam241** [comment]
I think the main problem is building geant4... that takes forever....

**Comment by gonuke** [comment]
And we build Geant4 four different times!  It times out on the first try.

**Comment by gonuke** [comment]
This may be a good reason to separate out some of the docker `RUN` lines, and build Geant4 before HDF5.  It is less likely to change in ways that matter, and we can do that locally, push the image and then let the autobuild skip the parts that are already built.

That said, perhaps we make a bunch of different images, each tagged with the info we we want and let dockerhub build them all?

TL;DR: if we want to leverage autobuilds on dockerhub, perhaps we need a different strategy.

**Comment by bam241** [comment]
Few though there and comments:
- the `hooks/build` strategy seems to have worked to allow multiple build using a single `Dockerfile` with different `ENV_VARIABLE`
- I am not sure how the 2h limits is used, is it 2h limits per job (i.e. `hooks/build` script), per image build or per RUN command....
- it seems that the `ci/dockercloud` hook replaced our `CI/CircleCI` hook in the test check :(, not sure if we can have both into the same repo... will look into it...

**Comment by bam241** [comment]
I am not sure what to do with this..., the dockeer build process is wait to long...

my suggestion would be to leave this on the side, maybe make a issue about this and merge this PR....

**Comment by bam241** [comment]
I am not sure how, but travis is back :)

**Comment by bam241** [comment]
discovered a bug in the existing `Dockerfile`, took me a couple of hours to fix it.
I am building and pushing fixed version of the docker container. 
I'll do my "final" push on this PR after :)

**Comment by bam241** [comment]
@gonuke dockerhub auto-build restriction to 2h runs, will force us to redesign our workflow concerning docker image generation and usage in CI. 
(I tried to push the image before hand but still takes a super long time on dockerhub...)
After discussing with @pshriwise we concluded that it is probably better to keep this PR simple, make an issue about dockerhub auto-generation, and redesign docker image workflow in a different PR.


@ljacobson64 @kkiesling is there anything blocking still this PR for you ? (I'll be happy to discuss/tackle those points if any remain.)

I'd like to keep this moving... @pshriwise's PR #674 is waiting for this to be finalized and merged.

**Comment by gonuke** [comment]
I think what I want to see is that the docker image generated by this process will result in an image in which testing is successful.  If you already pushed a new image based on this to dockerhub, maybe that already happened when the most recent commits were tested on Travis??  can you confirm?

**Comment by bam241** [comment]
Yes I pushed my last commit after having build and push the new docker images on `dockerhub`.

so the last set of Travis test were run on docker images generated with this docker image

**Comment by bam241** [comment]
todo:
- [x] move housekeeping the the beginning
- [x] stop CI if one test fail

File: docker/Dockerfile
Comment (bam241) [reply docker/Dockerfile:18]: I re-added those line for ubuntu 18.04 only as housekeeping is only done there.

File: docker/Dockerfile
Comment (bam241) [reply docker/Dockerfile:17]: I put it back in :)

**Comment by bam241** [comment]
@pshriwise @gonuke 
- I added a fast fail, which will stop CI jobs if one fails. (I try to with a DagMC in which I messed up the formatting, it fails and stoped everything [see](https://travis-ci.org/github/svalinn/DAGMC/builds/687020418))
- I reput housekeeping as the first jobs, so it should run first.

**Comment by bam241** [comment]
@pshriwise @gonuke 
- I added a fast fail, which will stop CI jobs if one fails. (I try to with a DagMC in which I messed up the formatting, it fails and stoped everything [see](https://travis-ci.org/github/svalinn/DAGMC/builds/687020418))
- I reput housekeeping as the first job, so it should run first.


@ljacobson64 I think I addressed all your comments, could you confirm ??

File: scripts/build_geant4.sh
Comment (ljacobson64) [inline scripts/build_geant4.sh]: If we're doing the shasum thing for Geant4, would it make sense to do it for HDF5 too?

File: .travis.yml
Comment (ljacobson64) [inline .travis.yml]: I don't think I understand why the scripts need to be in a separate directory from the Dockerfile. All of the scripts in this directory are CI-specific scripts that are only intended to be used in conjunction with the docker image.

**Comment by ljacobson64** [comment]
One final comment: I think it would be a good idea to rebase this PR against `develop` before merging. There have been several PRs merged since this one was first submitted, so we want to make sure these new PRs also play nice with the new CI.

File: scripts/build_geant4.sh
Comment (bam241) [reply scripts/build_geant4.sh]: fair point !

File: .travis.yml
Comment (bam241) [reply .travis.yml]: I think I got confuse because some of the file where `travis.xxx` and some just `build` I though that the build one could be used elsewhere.
@ljacobson64 What about we put all of them in a CI folder as you suggested (including the `Dockerfile`) and add a `docker.` prefix in the build scripts that are intended to be used in the docker building process.

Comment (ljacobson64) [reply .travis.yml]: I like that idea. So there'd be a folder called `CI` and it would contain these files

```
docker.build.dagmc.sh
docker.build.geant4.sh
docker.build.hdf5.sh
docker.build.moab.sh
docker.env.sh
Dockerfile
travis.housekeeping.sh
travis.install.sh
travis.tests.sh
update_docker.sh
```

Comment (bam241) [reply .travis.yml]: build_dagmc is only used in travis, so I named it `travis.`
still working on making sure update_docker still work before pushing it

Comment (bam241) [reply .travis.yml]: `env.sh` is used in both, so I kept it without a prefix

**Comment by bam241** [comment]
@ljacobson64 I added the suggestion about check_sum for hdf5 (and the moving/renaming of the scripts and Dockerfile)

the hdf5 shasum made me think about the version we are using, as when we use an other version than the system one, we only use the `1.10.4` shall we note specify it in the `env.sh` instead of having a environment variable that we set "manually" ?

File: .travis.yml
Comment (gonuke) [reply .travis.yml]: Sorry to be late here, but why `CI/travis.xxx` and `CI/docker.xxxx` and not `CI/travis/xxxx` and `CI/docker/xxxx`?

File: CI/docker.build_hdf5.sh
Comment (ljacobson64) [inline CI/docker.build_hdf5.sh]: `$hdf5_shasum` should be `$hdf5_1_10_4_shasum`

**Comment by bam241** [comment]
@ljacobson64 @gonuke I this I addressed your last comments :)

**Review by gonuke — APPROVED** [review]
Thanks @bam241

**Comment by bam241** [comment]
@ljacobson64 do you have any other comments ?

File: CI/update_docker.sh
Comment (ljacobson64) [inline CI/update_docker.sh:8]: Is this script now intended to be run from outside the `CI` directory?

Comment (bam241) [reply CI/update_docker.sh:8]: I though it would make more sense to run from the top directory

**Review by ljacobson64 — APPROVED** [review]

**Comment by ljacobson64** [comment]
Has the docker image on dockerhub already been updated? I'm curious about how the tests are passing if it hasn't. Unless the new docker image is the same as the old one despite having a new Dockerfile.

**Comment by bam241** [comment]
Yes the docker container have been updated/pushed after last dockerfile change.

**Review by gonuke — APPROVED** [review]
Thank @bam241 - LGTM

**Pull request merged**
